File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq;
        for(auto ele: stones){
            pq.push(ele);
        }

        while(pq.size()>1){
            int first  = pq.top(); pq.pop();
            int second = pq.top(); pq.pop();

            if(first!=second){
                pq.push((first-second));
            }
        }

        if(!pq.empty()) return pq.top();
        return 0;
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.py
import heapq
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        maxHeap = []
        for ele in stones:
            maxHeap.append(-ele)
        heapq.heapify(maxHeap)


        while(len(maxHeap)>1):
            first = -heapq.heappop(maxHeap)
            second = -heapq.heappop(maxHeap)

            if(first!=second):
                heapq.heappush(maxHeap,-(first-second))

        if maxHeap:
            return -maxHeap[0]

        return 0
            

        